lib/SIL/SILFunctionType.cpp

```diff
@@ -1791,8 +1791,94 @@ bool TypeConverter::requiresNewVTableEntry(SILDeclRef method) {
   return result;
 }
 
-bool TypeConverter::requiresNewVTableEntryUncached(SILDeclRef method) {
-  auto *decl = method.getDecl();
+// This check duplicates TypeConverter::checkForABIDifferences(),
+// but on AST types. The issue is we only want to introduce a new
+// vtable thunk if the AST type changes, but an abstraction change
+// is OK; we don't want a new entry if an @in parameter became
+// @guaranteed or whatever.
+static bool checkASTTypeForABIDifferences(CanType type1,
+                                          CanType type2) {
+  // Unwrap optionals, but remember that we did.
+  bool type1WasOptional = false;
+  bool type2WasOptional = false;
+  if (auto object = type1.getAnyOptionalObjectType()) {
+    type1WasOptional = true;
+    type1 = object;
+  }
+  if (auto object = type2.getAnyOptionalObjectType()) {
+    type2WasOptional = true;
+    type2 = object;
+  }
+
+  // Forcing IUOs always requires a new vtable entry.
+  if (type1WasOptional && !type2WasOptional)
+    return true;
+
+  // Except for the above case, we should not be making a value less optional.
+
+  // If we're introducing a level of optionality, only certain types are
+  // ABI-compatible -- check below.
+  bool optionalityChange = (!type1WasOptional && type2WasOptional);
+
+  // If the types are identical and there was no optionality change,
+  // we're done.
+  if (type1 == type2 && !optionalityChange)
+    return false;
+
+  // Classes, class-constrained archetypes, and pure-ObjC existential types
+  // all have single retainable pointer representation; optionality change
+  // is allowed.
+  if ((type1->mayHaveSuperclass() ||
+       type1->isObjCExistentialType()) &&
+      (type2->mayHaveSuperclass() ||
+       type2->isObjCExistentialType()))
+    return false;
+
+  // Class metatypes are ABI-compatible even under optionality change.
+  if (auto metaTy1 = dyn_cast<MetatypeType>(type1)) {
+    if (auto metaTy2 = dyn_cast<MetatypeType>(type2)) {
+      if (metaTy1.getInstanceType().getClassOrBoundGenericClass() &&
+          metaTy2.getInstanceType().getClassOrBoundGenericClass())
+        return false;
+    }
+  }
+
+  if (!optionalityChange) {
+    // Function parameters are ABI compatible if their differences are
+    // trivial.
+    if (auto fnTy1 = dyn_cast<AnyFunctionType>(type1)) {
+      if (auto fnTy2 = dyn_cast<AnyFunctionType>(type2)) {
+        return (
+            checkASTTypeForABIDifferences(fnTy2.getInput(), fnTy1.getInput()) ||
+            checkASTTypeForABIDifferences(fnTy1.getResult(), fnTy2.getResult()));
+      }
+    }
```

Comment:
For function types, you want to ensure the representations are the same too, since `@convention(block|c|swift)` aren't ABI compatible. I'm not sure whether we accept an override that changes convention, but it wouldn't be unreasonable to since the different-convention types are interconvertible.